From 3934c0eeb6b88e721ecc91b1cd4f293f6f0fac24 Mon Sep 17 00:00:00 2001
From: Andy Ross <andrew.j.ross@intel.com>
Date: Wed, 8 Apr 2015 10:05:23 -0700
Subject: [PATCH 1/2] atomisp_gmin_platform: Disable ACPI PM on camera devices

We don't use this support, but BIOS DSDTs inherit it from the windows
model and the default handling has introduced bugs like large sleeps
during suspend/resume and/or leaving rails on incorrectly following
suspend cycles.  Disable always.  If/when we need to support this
model in the driver we can do it intelligently via proper runtime PM.

Change-Id: Ifbe60b0b3f155fe8d3612b6999b52d20abd8f2dd
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-7324
Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
---
 arch/x86/platform/intel-mid/atomisp_gmin_platform.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
index 474d410..a907884 100644
--- a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
+++ b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
@@ -188,9 +188,18 @@ int atomisp_register_i2c_module(struct v4l2_subdev *subdev,
 	struct i2c_board_info *bi;
 	struct gmin_subdev *gs;
         struct i2c_client *client = v4l2_get_subdevdata(subdev);
+	struct acpi_device *adev;
 
 	dev_info(&client->dev, "register atomisp i2c module type %d\n", type);
 
+	/* The windows driver model (and thus most BIOSes by default)
+	 * uses ACPI runtime power management for camera devices, but
+	 * we don't.  Disable it, or else the rails will be needlessly
+	 * tickled during suspend/resume.  This has caused power and
+	 * performance issues on multiple devices. */
+	adev = ACPI_COMPANION(&client->dev);
+	if (adev)
+		adev->power.flags.power_resources = 0;
 
 	for (i=0; i < MAX_SUBDEVS; i++)
 		if (!pdata.subdevs[i].type)
-- 
1.9.1

